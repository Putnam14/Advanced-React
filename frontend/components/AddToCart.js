import { Component } from 'react';
import { Mutation } from 'react-apollo';
import PropTypes from 'prop-types';
import { ADD_TO_CART_MUTATION, CURRENT_USER_QUERY } from '../queries/index';

class AddToCart extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  };

  update = (proxy, payload) => {
    console.log('Updating!');
    console.log(payload);
    const newCartItem = payload.data.addToCart;
    const data = proxy.readQuery({ query: CURRENT_USER_QUERY });
    console.log(newCartItem);

    const existingIndex = data.me.cart.findIndex(cartItem => cartItem.id === newCartItem.id);
    if (existingIndex >= 0) {
      // already in cache, just replace it
      data.me.cart = [
        ...data.me.cart.slice(0, existingIndex),
        newCartItem,
        ...data.me.cart.slice(existingIndex + 1),
      ];
    } else {
      data.me.cart = [...data.me.cart, newCartItem];
    }
    proxy.writeQuery({ query: CURRENT_USER_QUERY, data });
  };

  render() {
    const { id } = this.props;
    return (
      <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ id }} update={this.update}>
        {(addToCart, { loading }) => (
          <button disabled={loading} onClick={addToCart}>
            ðŸ›’ Add{loading && 'ing'} To Cart
          </button>
        )}
      </Mutation>
    );
  }
}

export default AddToCart;
