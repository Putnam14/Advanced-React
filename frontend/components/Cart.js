import React from 'react';
import { Mutation, Query } from 'react-apollo';
import { adopt } from 'react-adopt';
import TakeMyMoney from './TakeMyMoney';
import formatMoney from '../lib/formatMoney';
import CartItem from './CartItem';
import { CURRENT_USER_QUERY, LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION } from '../queries/queries';
import calcTotalPrice from '../lib/calcTotalPrice';
import Error from './ErrorMessage';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';

const Composed = adopt({
  toggleCart: <Mutation mutation={TOGGLE_CART_MUTATION}>{() => {}}</Mutation>,
  localState: <Query query={LOCAL_STATE_QUERY}>{() => {}}</Query>,
  currentUser: (
    <Query query={CURRENT_USER_QUERY} data-test="cart">
      {() => {}}
    </Query>
  ),
});

const Cart = () => (
  <Composed>
    {({ toggleCart, localState, currentUser }) => {
      const { data: { me }, error, loading } = currentUser;
      if (loading) return <p>Loading...</p>;
      if (error) return <Error error={error} />;
      if (!me) return <p>Please Sign In!</p>;
      return (
        <CartStyles open={localState.data.cartOpen}>
          <header>
            <CloseButton title="close" onClick={toggleCart}>
              &times;
            </CloseButton>

            <Supreme>{me.name}'s Cart.</Supreme>
            <p>
              You have {me.cart.length} item{me.cart.length === 1 ? '' : 's'} in your cart.
            </p>
          </header>

          <ul>{me.cart.map(cartItem => <CartItem key={cartItem.id} cartItem={cartItem} />)}</ul>
          <footer>
            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
            <TakeMyMoney>
              <SickButton>Checkout</SickButton>
            </TakeMyMoney>
          </footer>
        </CartStyles>
      );
    }}
  </Composed>
);

export default Cart;
