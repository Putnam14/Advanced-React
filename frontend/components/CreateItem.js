import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import { CREATE_ITEM_MUTATION, ALL_ITEMS_QUERY } from '../queries/queries';
import Error from './ErrorMessage';
import Form from './styles/Form';
import PropTypes from 'prop-types';
import Router from 'next/router';
import formatMoney from '../lib/formatMoney';

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  };

  handleChange = e => {
    const { name, value, type } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  uploadFile = async e => {
    this.setState({ loading: true });
    const files = e.currentTarget.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');

    // use the file endpoint
    const res = await fetch('https://api.cloudinary.com/v1_1/wesbos/image/upload', {
      method: 'POST',
      body: data,
    });
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
      loading: false,
    });
  };

  // TODO: Problem here - if I refetch page 1, page 2 to infinity are still out of date. I need to refetch all the pages somehow? Or use @connection
  render() {
    return (
      <Mutation
        mutation={CREATE_ITEM_MUTATION}
        variables={this.state}
        // TODO - update all items
        // refetchQueries={[{ query: ALL_ITEMS_QUERY }]}
      >
        {(createItem, { loading, error }) => (
          <Form
            data-test
            onSubmit={async e => {
              e.preventDefault();
              const { data: { createItem: item } } = await createItem();
              Router.push({
                pathname: `/item`,
                query: { id: item.id },
              });
            }}
          >
            <h2>Sell an Item.</h2>
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label>
                Image
                <input required onChange={this.uploadFile} type="file" accept=".png, .jpg, .jpeg" />
                {this.state.image ? (
                  <img src={this.state.image} width="100" alt={this.state.title} />
                ) : null}
              </label>
              <label>
                Title
                <input
                  required
                  value={this.state.title}
                  onChange={this.handleChange}
                  type="text"
                  name="title"
                  id="title"
                  placeholder="Title"
                />
              </label>
              <label>
                Price {this.state.price && formatMoney(this.state.price)}
                <input
                  required
                  type="number"
                  id="price"
                  name="price"
                  min="0"
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>
              <textarea
                id="description"
                required
                name="description"
                value={this.state.description}
                onChange={this.handleChange}
                placeholder="The desc for this item"
              />
              <button disabled={this.state.loading} type="submit">
                Submit
              </button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
