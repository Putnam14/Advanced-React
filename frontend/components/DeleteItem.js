import React from 'react';
import { Mutation } from 'react-apollo';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { withRouter } from 'next/router';
import { ALL_ITEMS_QUERY } from './Items';
import { perPage } from '../config';

const DELETE_ITEM_MUTATION = gql`
  mutation deleteItem($id: ID!) {
    deleteItem(id: $id) {
      id
      title
      description
    }
  }
`;

class DeleteItem extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  };

  update = (cache, payload) => {
    const deletedItem = payload.data.deleteItem;
    const { page = 1 } = this.props.router.query;
    const skip = page * perPage - perPage;
    const variables = { skip };
    // TODO Expose Page Number here
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY, variables });
    console.log(data.items);
    // filter this one out
    data.items = data.items.filter(item => item.id !== deletedItem.id);
    // write the data back to the cache
    console.log(data.items);
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data, variables });
  };

  render() {
    console.log(this.props.router.query);
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: this.props.id }}
        refetchQueries={[
          {
            query: ALL_ITEMS_QUERY,
            variables: { skip: this.props.router.query.page * perPage - perPage },
          },
        ]}
        update={this.update}
      >
        {(removeItem, { error }) => (
          <button
            onClick={() => {
              if (confirm('Are you sure you want to delete this item?')) {
                removeItem();
              }
            }}
          >
            {error ? error.message : 'Ã— Delete Item'}
          </button>
        )}
      </Mutation>
    );
  }
}

export default withRouter(DeleteItem);
export { DELETE_ITEM_MUTATION };
