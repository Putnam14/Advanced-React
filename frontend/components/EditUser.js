import React from 'react';
import { Query, Mutation } from 'react-apollo';
import Form from './styles/Form';
import { CURRENT_USER_QUERY, UPDATE_USER_MUTATION } from '../queries/queries';
import Error from './ErrorMessage';

class EditUser extends React.Component {
  state = {
    changes: {},
  };

  handleChange = e => {
    const { name, value } = e.target;
    const changes = {
      ...this.state.changes,
      [name]: value,
    };
    this.setState({ changes });
  };

  handleSubmit = async (e, updateUser) => {
    e.preventDefault();
    // only submit if there are real changes
    if (Object.keys(this.state.changes).length === 0) return;
    await updateUser();
    this.setState({ changes: {} });
  };

  render() {
    return (
      <Query query={CURRENT_USER_QUERY}>
        {({ data: { me }, loading }) => {
          if (loading) return <p>Loading...</p>;
          return (
            <Mutation
              mutation={UPDATE_USER_MUTATION}
              refetchQueries={[{ query: CURRENT_USER_QUERY }]}
              variables={this.state.changes}
            >
              {(updateUser, { loading, error, called, data }) => (
                <Form onSubmit={e => this.handleSubmit(e, updateUser)}>
                  <Error error={error} />
                  <fieldset disabled={loading} aria-busy={loading}>
                    {called && !error && <p data-test="updated">Updated!</p>}
                    <label htmlFor="name">
                      Name:
                      <input
                        type="text"
                        name="name"
                        defaultValue={me.name}
                        onChange={this.handleChange}
                      />
                    </label>
                    <button type="submit">Update</button>
                    <strong>me:</strong>
                    <pre>{JSON.stringify(me.name)}</pre>
                    <strong>Change:</strong>
                    <pre data-test="change">{JSON.stringify(this.state.changes)}</pre>
                  </fieldset>
                </Form>
              )}
            </Mutation>
          );
        }}
      </Query>
    );
  }
}

export default EditUser;
