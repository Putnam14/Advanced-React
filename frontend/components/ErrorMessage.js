import styled from 'styled-components';
import React from 'react';

import PropTypes from 'prop-types';

const StyledError = styled.div`
  padding: 2rem;
  background: white;
  margin: 2rem 0;
  border: 1px solid rgba(0, 0, 0, 0.05);
  border-left: 5px solid red;
  p {
    margin: 0;
    font-weight: 100;
  }
  strong {
    margin-right: 1rem;
  }
`;

const DisplayError = ({ error, refetch }) => {
  if (!error || !error.message) return null;
  if (error.networkError && error.networkError.result && error.networkError.result.errors.length) {
    return error.networkError.result.errors.map((error, i) => (
      <StyledError key={i}>
        <p data-test="graphql-error">
          <strong>Shoot!</strong>
          {error.message.replace('GraphQL error: ', '')}
        </p>
      </StyledError>
    ));
  }
  return (
    <StyledError>
      <p data-test="graphql-error">
        <strong>Shoot!</strong>
        {error.message.replace('GraphQL error: ', '')}
        <button onClick={refetch}>Try Again</button>
      </p>
    </StyledError>
  );
};

DisplayError.propTypes = {
  error: PropTypes.object,
};

export default DisplayError;
