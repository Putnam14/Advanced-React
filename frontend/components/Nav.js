import React, { Fragment } from 'react';
import Link from 'next/link';
import { Query, ApolloConsumer } from 'react-apollo';
import User from './User';
import CartCount from './CartCount';
import Signout from './Signout';
import NavStyles from './styles/NavStyles';

class Nav extends React.Component {
  render() {
    // below we set the fetchPolicy to network only so it forces re-fetch on the server
    return (
      <User>
        {({ data: { me }, error }) => (
          <NavStyles data-test="nav">
            <Link href="/items">
              <a>Shop</a>
            </Link>
            <Link href="/sell">
              <a>Sell</a>
            </Link>

            {!me && (
              <Link href="/signup">
                <a>Sign In</a>
              </Link>
            )}

            {me && (
              <Fragment>
                <Link href="/orders">
                  <a>Orders</a>
                </Link>
                <Link href="/me">
                  <a>My Account</a>
                </Link>
                <Signout />
                <ApolloConsumer>
                  {cache => (
                    <button onClick={() => cache.writeData({ data: { cartOpen: true } })}>
                      My Cart
                      <CartCount
                        className="cart-count"
                        count={me.cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)}
                      />
                    </button>
                  )}
                </ApolloConsumer>
              </Fragment>
            )}
          </NavStyles>
        )}
      </User>
    );
  }
}

export default Nav;
