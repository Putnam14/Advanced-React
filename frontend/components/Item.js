import React from 'react';
import styled from 'styled-components';
import Link from 'next/link';
import Title from './styles/Title';
import AddToCart from './AddToCart';
import DeleteItem from './DeleteItem';
import formatMoney from '../lib/formatMoney';
import PropTypes from 'prop-types';

const Item = styled.div`
  background: white;
  border: 1px solid ${props => props.theme.offWhite};
  box-shadow: ${props => props.theme.bs};
  position: relative;
  display: grid;
  align-content: start;
  grid-auto-rows: fit-content;
  img {
    width: 100%;
  }
  p {
    font-size: 14px;
    line-height: 2;
    font-weight: 600;
    padding: 0 3rem;
    font-size: 1.5rem;
  }
  .buttonList {
    display: grid;
    border-top: 1px solid ${props => props.theme.lightgrey};
    grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
    grid-gap: 1px;
    background: ${props => props.theme.lightgrey};
    align-self: end;
    & > * {
      background: white;
      border: 0;
      font-size: 1rem;
      padding: 1rem;
    }
  }
`;

const PriceTag = styled.span`
  background: ${props => props.theme.red};
  transform: rotate(3deg);
  color: white;
  font-weight: 600;
  padding: 5px;
  line-height: 1;
  font-size: 3rem;
  display: inline-block;
  position: absolute;
  top: -3px;
  right: -3px;
`;

class ItemComponent extends React.Component {
  static propTypes = {
    item: PropTypes.object.isRequired,
  };

  render() {
    const item = this.props.item;
    return (
      <Item key={item.id}>
        {item.image ? <img src={item.image} alt={item.title} /> : null}
        <Title>
          <Link
            href={{
              pathname: '/item',
              query: { id: item.id },
            }}
          >
            <a>{item.title}</a>
          </Link>
        </Title>

        <PriceTag>{formatMoney(item.price)}</PriceTag>

        <p>{item.description}</p>

        <div className="buttonList">
          <Link
            href={{
              pathname: '/update',
              query: { id: item.id },
            }}
          >
            <a>Edit ✏️</a>
          </Link>
          <AddToCart id={item.id} />
          <DeleteItem id={item.id} />
        </div>
      </Item>
    );
  }
}

export default ItemComponent;
