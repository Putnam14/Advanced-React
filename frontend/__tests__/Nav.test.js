import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import Router from 'next/router';
import Nav from '../components/Nav';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeUser } from '../lib/testUtils';
import { CURRENT_USER_QUERY } from '../components/User';

// Mock the router
Router.router = { push() {} };

const loggedOutMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    delay: 0,
    result: {
      data: {
        me: null,
      },
    },
  },
];

const loggedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        what: 'yah',
        me: {
          ...fakeUser(),
          orders: [],
          cart: [],
        },
      },
    },
  },
];

describe('<Nav></Nav>', () => {
  it('Renders minimal nav when logged out', () => {
    const wrapper = mount(
      <MockedProvider mocks={loggedOutMocks}>
        <Nav />
      </MockedProvider>
    );
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(toJSON(nav)).toMatchSnapshot();
  });

  it('renders full nav when logged in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={loggedInMocks}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const nav = wrapper.find('ul[data-test="nav"]');
    // account link
    const accountLink = wrapper.find('a[href="/me"]');
    expect(accountLink.length).toBe(1);
  });
});
