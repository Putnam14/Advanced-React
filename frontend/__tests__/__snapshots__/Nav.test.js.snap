// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav></Nav> Renders minimal nav when logged out 1`] = `
<ul
  className="sc-dnqmqq kxZizK"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/sell"
  >
    <a
      href="/sell"
      onClick={[Function]}
    >
      Sell
    </a>
  </Link>
  <Link
    href="/signup"
  >
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign In
    </a>
  </Link>
</ul>
`;
